tests/feature/PermissionsTest.php

```diff
@@ -0,0 +1,242 @@
+
+		$result->assertStatus(403);
+		$result->assertJSONFragment(['error' => lang('Permits.notPermitted')]);
+	}
+
+	public function testAuthenticatedAddOnlyEmptyFile()
+	{
+		$this->setMode(04664);
+		$this->login($this->admin->id);
+
+		$result = $this->withSession()
+                 		->withHeaders(['X-Requested-With' => 'XMLHttpRequest'])
+		                ->post('files/upload');
+		$result->assertStatus(400);
+	}
+
+	public function testAuthenticatedAddOnlyWithInvalidFile()
+	{
+		$this->setMode(04664);
+		$this->login($this->admin->id);
+
+		$_FILES = [
+			'file' => [
+				'name'     => 'someFile.txt',
+				'type'     => 'text/plain',
+				'size'     => '124',
+				'tmp_name' => '/tmp/myTempFile.txt',
+				'error'    => 0,
+			],
+		];
+
+		$result = $this->withSession()
+                 	   ->withHeaders(['X-Requested-With' => 'XMLHttpRequest'])
+		               ->post('files/upload');
+		$result->assertSee('The file uploaded with success.(0)');
+	}
+
+	public function testAuthenticatedAddOnlyWithValidFile()
+	{
+		$this->setMode(04664);
+		$this->login($this->admin->id);
+
+		$_FILES = [
+			'file' => [
+				'name'     => 'file.txt',
+				'type'     => 'text/plain',
+				'size'     => '33',
+				'tmp_name' => 'tests/_support/vfs/file.txt',
+				'error'    => 0,
+			],
+		];
+
+		p\redefine(UploadedFile::class . '::isValid', function() {
+			return true;
+		});
+
+		p\redefine(File::class . '::move', function($args) {
+			return new File('tests/_support/vfs/file.txt');
+		});
+
+		$modelClass = get_class($this->model);
+		p\redefine($modelClass . '::createFromPath', function($args) {
+			return fake(FileFaker::class);
+		});
+
+		$result = $this->withSession()
+                 	   ->withHeaders(['X-Requested-With' => 'XMLHttpRequest'])
+					   ->post('files/upload');
+
+		$this->assertEquals('',	$result->response->getBody());
+	}
+
+	public function testProctorListsAllFiles()
+	{
+		$this->setMode(00660);
+		$this->login($this->proctor->id);
+
+		$result = $this->withSession()->get('files');
+		$result->assertStatus(200);
+
+		$files = $this->model->getForUser($this->admin->id);
+		$result->assertSee($files[0]->filename);
+	}
+
+	public function testAuthenticatedListOwnOnly()
+	{
+		$this->setMode(00660);
+		$this->login($this->proctor->id);
+
+		$fileOwnByProctor = fake(FileFaker::class);
+		model('FileModel')->addToUser($fileOwnByProctor->id, $this->proctor->id);
+
+		$fileOwnByProctor2 = fake(FileFaker::class);
+		model('FileModel')->addToUser($fileOwnByProctor2->id, $this->proctor->id);
+
+		$fileOwnByAdmin = fake(FileFaker::class);
+		model('FileModel')->addToUser($fileOwnByAdmin->id, $this->admin->id);
+
+		$result = $this->withSession()->get('files/user/' . $this->proctor->id);
+		$result->assertStatus(200);
+
+		$files = $this->model->getForUser($this->proctor->id);
+		$result->assertSee($fileOwnByProctor->filename);
+		$result->assertSee($fileOwnByProctor2->filename);
```

Comment:
A user visiting `files/user/{their_own_id}` should only see their own files. Something is wrong if another user's file is showing up.

Reply:
Yes, both 2 line are for addition to proctor user, in next line, there is :
 ```php
$result->assertDontSee($fileOwnByAdmin->filename)
````
 for **admin** that proctor can't see admin file, is that already correct ?

Reply:
Ohhhh yes you are right, I misread that as "file owned by proctor 2 {a different user}". Looks good!